instrumentation/resque/lib/opentelemetry/instrumentation/resque/instrumentation.rb

```diff
@@ -18,6 +18,28 @@ class Instrumentation < OpenTelemetry::Instrumentation::Base
           defined?(::Resque)
         end
 
+        ## Supported configuration keys for the install config hash:
+        #
+        # force_flush: when `true`, all completed spans will be synchronously flushed
+        #   at the end of a job's execution (default: `false`). You will likely wish to
+        #   enable this option for job systems that fork worker processes such as Resque.
```

Comment:
It's a little odd to read `such as Resque` in instrumentation specifically for Resque. I *think* this option is useful when running Resque in a certain mode, but it's been a couple years since I was intimate with Resque. Can you clarify this a little?

Reply:
Good catch, this is what I get for adopting and not adapting the ActiveJob implementation. 

Updated: https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/117/commits/88c18b457edb641e672cb30c948383684edb7794

Reply:
As for forking, as far as I know, this is the default behaviour for Resque **[Fork Documentation Link](https://github.com/resque/resque/#forking)**. While the documentation does begin with `"On certain platforms..."`, I am only familiar with using Resque when workers fork off child processes.

Reply:
I could append something like this to the `:force_flush` comment, what do you think?

```ruby
"... You will likely wish to enable this option if `Resque::Worker#fork_per_job?` is enabled."
```

Reply:
I guess we're quibbling over semantics, but if it's `you will likely wish` then it makes me wonder "why isn't it the default?" (I know why, I'm just thinking as I would if this were my first time seeing it).

So: yes, I think adding that to the doc is wise, I would say "You may wish" instead.

Reply:
Agreed!

Reply:
What if we asked the job instead?

The job class seems to have a [`worker` accessor](https://github.com/resque/resque/blob/8ea2e4b41bae7cf4c1242b81b25eaa39b9ccaf86/lib/resque/job.rb#L67), and while there is no explicit accessor, the [`fork_per_job?` method on the worker](https://github.com/resque/resque/blob/8ea2e4b41bae7cf4c1242b81b25eaa39b9ccaf86/lib/resque/worker.rb#L792-L795) is not private and memoizes the result ... so it should be low-impact to check on every job.

I could envision the code looking something like this:

```ruby
if worker.fork_per_job?
  OpenTelemetry.tracer_provider.force_flush unless !!!config[:force_flush]
end
```

... but that only works, of course, if `config[:force_flush]` can be left in an _unset_ state by default. If we could do that, then the "do we flush" truth table looks like:

|                                 | `fork_per_job? == true` | `fork_per_job? == false` |
|---------------------------------|-------------------------|--------------------------|
| `config[:force_flush] == true`  | ✅                      | ✅                       |
| `config[:force_flush] == false` | 🚫                      | 🚫                       |
| `config[:force_flush] == nil`   | ✅                      | 🚫                       |

... I don't know if allowing this option to be `nil` is something we want. But it would get around the brittleness of trying to check this at startup, and would also avoid force-flushing needlessly on platforms where that isn't happening.

Reply:
If we go down that path, I'd recommend turning the option into an enum, e.g.:
```ruby
        option :force_flush,       default: :ask_the_job,  validate: %I[ask_the_job always never]
```
As noted above:
> Ideally, we'd be careful here not to break forks of Resque.

We should check `Hedwig`, for example, to make sure we don't break it. (We should also check whether we're using this instrumentation with `Hedwig` - if it has its own instrumentation, this is a moot point.)

Reply:
I checked. We have services using the upstream `Resque` with this instrumentation. `Hedwig` does not use this instrumentation - it has its own first-party instrumentation using OpenTelemetry. So I think this is safe from Shopify's perspective.

Reply:
I like the direction this conversation has taken and I am in favour of the enum suggestion made by @fbogsany. If we can land on a consensus within the maintainer group I'd be happy to push a commit that reflects this suggestion.

Reply:
For the sake of conversation, I'd pushed [b07e9ed](https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/117/commits/b07e9ed0bb4d1a933bb3d7d4dc1857285934a07a).

Reply:
Not too much urgency here, but I would like to get this merged soonish as it will unblock enabling the resque instrumentation on our side (and avoid a monkey-patch).